🌱  Bump Go 1.24

sivchari opened the pull request:
<!-- Thanks for sending a pull request! Here are some tips for you:
    1. If this is your first time, please read our contributor guidelines: https://github.com/kubernetes-sigs/cluster-api/blob/main/CONTRIBUTING.md#contributing-a-patch and developer guide https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/book/src/developer/getting-started.md

    2. Please add an icon to the title of this PR (see https://sigs.k8s.io/cluster-api/CONTRIBUTING.md#contributing-a-patch), and delete this line and similar ones
    the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🌱 (:seedling:, minor or other) 
-->

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

<!-- 
Please label this pull request according to what area(s) you are addressing. For reference on PR/issue labels, see: https://github.com/kubernetes-sigs/cluster-api/labels?q=area+

Area example:
/area runtime-sdk
-->

Part of #11642

/area dependency

sivchari commented:
> Error: can't load config: the Go language version (go1.23) used to build golangci-lint is lower than the targeted Go version (1.24)
  Failed executing command with error: can't load config: the Go language version (go1.23) used to build golangci-lint is lower than the targeted Go version (1.24)

We should merge https://github.com/kubernetes-sigs/cluster-api/pull/12088 at first since the golangci-lint version that Go1.24 is available is only golangci-lint v2.

sivchari commented on Makefile:
```diff
@@ -1,4 +1,4 @@
-# Copyright 2018 The Kubernetes Authors.
+:# Copyright 2018 The Kubernetes Authors.
```
Good catch! fixed.

fabriziopandini commented:
/hold

We usually use the same go version used by K8s version we are importing, and we import the same K8s version used by the controller runtime version we are using.
(so most probably we should stick to go 1.23 until we bump to CR 0.21)

sivchari commented:
Thanks!
Other PRs about dependency could be merged, when new CR is published.

sbueringer commented:
I would like to do the CR bump. I'm usually using CAPI as a final verification for CR *before* the CR release is published. I will coordinate with you that this PR is merged right before.

Can you please rebase this PR

sivchari commented:
@sbueringer 
rebased it.

sivchari commented on .golangci.yml:
```diff
@@ -47,6 +47,9 @@ linters:
     - usetesting # report function to be replace by testing
     - whitespace # unnecessary newlines
+  disable:
+    # TODO: It will be dropped when the Go version migration is done.
```
usetesting is automatically enabled when Go 1.24 is used. This linter diagnose the calls that can be replaced by functions defined in testing package. e.g. os.Setenv -> testing.Setenv.
It is what I'm going to do after the migration to Go1.24 since the diff is very large. So I disable it once, then would tackle it later on another PR.

sbueringer replied:
Sounds good! Maybe open a small issue to track it

sivchari replied:
I opened https://github.com/kubernetes-sigs/cluster-api/issues/12179 👍

sbueringer replied:
Thx!

sivchari commented on .golangci.yml:
```diff
@@ -224,6 +227,10 @@ linters:
+    govet:
+      disable:
+        # TODO: It will be dropped when the Go version migration is done.
+        - printf # printf is introduced since Go 1.24
```
Same here.

sivchari replied:
I deal with it in addition.

sivchari commented:
This failure is led by Go 1.24 since the go test run the vet implicitly before run test. Since 1.24, printf linter is called every time, then I disabled it on golanci.yml to deal with these issues on another PR. But it seems to deal with it too including with this PR.

sivchari commented:
```
=== Symbol Results ===

Vulnerability #1: GO-2025-3563
    Request smuggling due to acceptance of invalid chunked data in net/http
  More info: https://pkg.go.dev/vuln/GO-2025-3563
  Standard library
    Found in: net/http/internal@go1.24
    Fixed in: net/http/internal@go1.24.2
    Example traces found:
      #1: cmd/clusterctl/client/repository/repository_gitlab.go:184:28: repository.gitLabRepository.GetFile calls io.ReadAll, which eventually calls internal.chunkedReader.Read

Your code is affected by 1 vulnerability from the Go standard library.
This scan found no other vulnerabilities in packages you import or modules you
require.
Use '-show verbose' for more details.
=== Symbol Results ===

No vulnerabilities found.

Your code is affected by 0 vulnerabilities.
This scan also found 1 vulnerability in packages you import and 0
vulnerabilities in modules you require, but your code doesn't appear to call
these vulnerabilities.
Use '-show verbose' for more details.
=== Symbol Results ===

Vulnerability #1: GO-2025-3563
    Request smuggling due to acceptance of invalid chunked data in net/http
  More info: https://pkg.go.dev/vuln/GO-2025-3563
  Standard library
    Found in: net/http/internal@go1.24
    Fixed in: net/http/internal@go1.24.2
    Example traces found:
      #1: infrastructure/container/docker.go:130:24: container.dockerRuntime.PullContainerImage calls io.ReadAll, which eventually calls internal.chunkedReader.Read

Your code is affected by 1 vulnerability from the Go standard library.
This scan found no other vulnerabilities in packages you import or modules you
require.
Use '-show verbose' for more details.
make: *** [verify-govulncheck] Error 1
```

We met the vulnerability, then I decide to upgrade from go1.24.0 to go1.24.2.

sivchari commented:
/retest

sbueringer commented:
/lgtm

so far

@sivchari Can you please rebase?

Btw, I hope I get the CR release/PR done in the next 1-2 weeks

sivchari commented:
Thx! rebased.

sbueringer commented:
Thx!

/lgtm

Waiting with merge until we can be sure that we'll get a CR v0.21.0 release ahead of the CAPI alpha we want to do in a few weeks

sbueringer commented on .golangci-kal.yml:
```diff
@@ -1,6 +1,6 @@
 version: "2"
```
Can you check if the images referenced in cloudbuild.yaml & cloudbuild-nightly.yaml already have Go 1.24?

Otherwise we have to bump the images (and potentially produce new versions of these images if there are none available with Go 1.24)

sbueringer replied:
@sivchari We have to do ~ this for 1.24: https://github.com/kubernetes/test-infra/pull/34159

Do you want to take that over?

chrischdi replied:
gcr.io/k8s-staging-test-infra/gcb-docker-gcloud@sha256:4e830b673791d5595719bc6c4ca62dce3746b4e20d749e45004254bc6ef0a140 still contains 1.23

chrischdi replied:
```
❯ cat usr/local/go/VERSION
go1.23.4
time 2024-11-27T20:27:20Z
```

(downloaded and unpacked the image + the above command)

chrischdi replied:
Prior art for bumping the gcb image:

https://github.com/kubernetes/test-infra/pull/34159

sivchari replied:
I'll take over it 👍

sivchari replied:
Pushed!

chrischdi commented:
lgtm, besides google cloud build image bump

chrischdi commented:
Last open point: update the cloudbuild yaml files to use
`gcr.io/k8s-staging-test-infra/gcb-docker-gcloud@sha256:fe4041d57c711070436d3c5b7702534b7158f464e9772aa9e333a5e0e24fd854` / v20250513-9264efb079 which contains go 1.24 🎉

sbueringer commented on cloudbuild.yaml:
```diff
@@ -5,7 +5,7 @@ options:
   machineType: 'E2_HIGHCPU_8'
 steps:
-  - name: 'gcr.io/k8s-staging-test-infra/gcb-docker-gcloud@sha256:4e830b673791d5595719bc6c4ca62dce3746b4e20d749e45004254bc6ef0a140' # v20250116-2a05ea7e3d
+  - name: 'gcr.io/k8s-staging-test-infra/gcb-docker-gcloud@sha256:fe4041d57c711070436d3c5b7702534b7158f464e9772aa9e333a5e0e24fd854' # v20250513-9264efb079
```
I cannot pull this image, can you?

```
docker pull gcr.io/k8s-staging-test-infra/gcb-docker-gcloud@sha256:fe4041d57c711070436d3c5b7702534b7158f464e9772aa9e333a5e0e24fd854
Error response from daemon: manifest for gcr.io/k8s-staging-test-infra/gcb-docker-gcloud@sha256:fe4041d57c711070436d3c5b7702534b7158f464e9772aa9e333a5e0e24fd854 not found: manifest unknown: Requested entity was not found.
```

sbueringer replied:
I think this sha works `sha256:63840f133e0dfeea0af9ef391210da7fab9d2676172e2967fccab0cd6110c4e7`

sivchari replied:
Thx! I fixed it.

sbueringer commented:
> We met the vulnerability, then I decide to upgrade from go1.24.0 to go1.24.2.

Nice catch btw.

I discussed this with Christian and Fabrizio. Given that we are going to merge the CR bump ahead of the next release anyway we would like to merge the current PR as soon as it's ready (last finding should be: https://github.com/kubernetes-sigs/cluster-api/pull/12128#discussion_r2089229892)

sbueringer commented:
/hold cancel
/lgtm
/approve

(double checked the gcb-docker-gcloud images, now all good)
